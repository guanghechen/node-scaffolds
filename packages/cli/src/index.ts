export * from './util/ansi'
export * from './util/command-exists'
export * from './util/config'
export * from './util/merge'
export * from './util/option'
export * from './util/stdin'
export * from './types'
