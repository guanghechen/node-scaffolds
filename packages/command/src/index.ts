export * from './command-top'
export * from './command'
export * from './definition/command'
export * from './definition/option'
export * from './util'
