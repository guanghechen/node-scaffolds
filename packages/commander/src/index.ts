export * from './command'
export * from './main'
export * from './sub'
export * from './top'
export * from './util'

export * from '@guanghechen/cli'
