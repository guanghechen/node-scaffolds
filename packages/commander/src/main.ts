import type { ICommandConfigurationOptions } from '@guanghechen/cli'
import type { CommandOptions } from 'commander'
import type { Command } from './command'

// Process main command.
export type IMainCommandProcessor<O extends ICommandConfigurationOptions> = (
  options: O,
) => void | Promise<void>

// Create main command
export type IMainCommandCreator<O extends ICommandConfigurationOptions> = (
  handle?: IMainCommandProcessor<O>,
) => Command

// Mount main command
export type IMainCommandMounter = (parentCommand: Command, opts?: CommandOptions) => void

// Execute main command.
export type IMainCommandExecutor<V = void> = (args: string[]) => Promise<V>

/**
 * Create main command mounter
 *
 * @param create  main command creator
 * @param handle  main command processor
 */
export function createMainCommandMounter<O extends ICommandConfigurationOptions>(
  create: IMainCommandCreator<O>,
  handle: IMainCommandProcessor<O>,
): IMainCommandMounter {
  return (program: Command, opts?: CommandOptions): void => {
    const command = create(handle)
    if (command.name().length <= 0) {
      command.name('__main__')
    }
    program.addCommand(command, { ...opts, isDefault: true })
  }
}

/**
 * Create main command executor
 *
 * @param create  main command creator
 * @param handle  main command processor
 */
export function createMainCommandExecutor<O extends ICommandConfigurationOptions>(
  create: IMainCommandCreator<O>,
  handle: IMainCommandProcessor<O>,
): IMainCommandExecutor {
  return (args: string[]): Promise<void> => {
    return new Promise(resolve => {
      const wrappedHandler = async (options: O): Promise<void> => {
        await handle(options)
        resolve()
      }

      const command = create(wrappedHandler)
      command.parse(args)
    })
  }
}
