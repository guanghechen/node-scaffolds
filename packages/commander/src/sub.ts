import type { ICommandConfigurationOptions } from '@guanghechen/cli'
import type { CommandOptions } from 'commander'
import type { Command } from './command'

export interface ISubCommandOptions extends ICommandConfigurationOptions {
  // nothing yet.
}

export interface ISubCommandProcessor<O extends ISubCommandOptions> {
  /**
   * Process the command arguments and options.
   * @param args      Command arguments.
   * @param options   Command options.
   */
  process(args: string[], options: O): Promise<void>
}

export interface ISubCommand<O extends ISubCommandOptions> extends ISubCommandProcessor<O> {
  /**
   * Sub command name.
   */
  readonly subCommandName: string

  /**
   * Sub command alias.
   */
  readonly aliases: string[]

  /**
   * Create an commander instance.
   * @param processor
   */
  command(processor: ISubCommandProcessor<O>): Command

  /**
   * Create a sub-command instance and execute it immediately.
   * @param args      Command arguments.
   * @param options   Command options.
   */
  execute(parentCommand: Command, rawArgs: string[]): Promise<void>

  /**
   * Create a sub-command instance and mount it to the parentCommand.
   * @param parentCommand
   * @param opts
   */
  mount(parentCommand: Command, opts: CommandOptions | undefined): void

  /**
   * Process the command options and rawArgs.
   * @param args      Command arguments.
   * @param options   Command options.
   */
  process(args: string[], options: O): Promise<void>

  /**
   * Resolve a processor for process the command.
   * @param args
   * @param options
   */
  resolveProcessor(args: string[], options: O): Promise<ISubCommandProcessor<O>>
}

export abstract class SubCommand<O extends ISubCommandOptions> implements ISubCommand<O> {
  public abstract subCommandName: string
  public abstract aliases: string[]

  public abstract command(processor: ISubCommandProcessor<O>): Command

  public mount(parentCommand: Command, opts: CommandOptions | undefined): void {
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    const processor: ISubCommandProcessor<O> = this
    const command: Command = this.command(processor)
    parentCommand.addCommand(command, opts)
  }

  public execute(parentCommand: Command, rawArgs: string[]): Promise<void> {
    return new Promise((resolve, reject) => {
      const processor: ISubCommandProcessor<O> = {
        process: async (args: string[], options: O): Promise<void> => {
          try {
            await this.process(args, options)
            resolve()
          } catch (error) {
            reject(error)
          }
        },
      }

      const command = this.command(processor)
      parentCommand.addCommand(command)
      parentCommand.parse(rawArgs)
    })
  }

  public async process(args: string[], options: O): Promise<void> {
    const processor: ISubCommandProcessor<O> = await this.resolveProcessor(args, options)
    await processor.process(args, options)
  }

  public abstract resolveProcessor(args: string[], options: O): Promise<ISubCommandProcessor<O>>
}
