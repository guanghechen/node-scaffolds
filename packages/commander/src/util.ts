import { safeExec } from '@guanghechen/exec'
import type { IReporter } from '@guanghechen/reporter.types'
import select from '@inquirer/select'
import commandExists from 'command-exists'

// Check if the git installed.
export const hasGitInstalled = (): boolean => commandExists.sync('git')

export interface IInstallDependenciesParams {
  readonly cwd: string
  readonly plopBypass: string[]
  readonly reporter?: IReporter
}

/**
 * Run `npm/yarn install` to Install node.js dependencies
 */
export async function installDependencies(params: IInstallDependenciesParams): Promise<void> {
  const { cwd, plopBypass, reporter } = params
  const hasYarnInstalled: boolean = commandExists.sync('yarn')

  /**
   * If neither yarn nor npm is installed, this operation will be skipped
   */
  if (!hasYarnInstalled) {
    const hasNpmInstalled: boolean = commandExists.sync('npm')
    if (!hasNpmInstalled) return
  }

  let npmScript: string
  if (plopBypass.length > 0) {
    npmScript = plopBypass.shift()!
  } else {
    npmScript = await select({
      message: 'npm or yarn?',
      choices: ['npm', 'yarn', 'skip'],
      default: hasYarnInstalled ? 'yarn' : 'npm',
    })
  }

  reporter?.debug?.('npmScript:', npmScript)

  // skip installing dependencies
  if (npmScript === 'skip') return

  // install dependencies
  await safeExec({ from: 'installDependencies', cmd: npmScript, args: ['install'], cwd, reporter })
}
