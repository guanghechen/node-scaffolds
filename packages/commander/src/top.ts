import { Command } from './command'

/**
 * Create top command
 * @param commandName
 * @param version
 */
export function createTopCommand(commandName: string, version: string): Command {
  const program = new Command()

  program
    .storeOptionsAsProperties(false) //
    .version(version)
    .name(commandName)

    // Register logger (reporter) related options.
    .option('--log-encoding <encoding>', 'Encoding of log file.')
    .option('--log-filepath <filepath>', 'Path which the log file is located.')
    .option('--log-level <level>', 'Log level.')
    .option('--log-name <name>', 'Logger name.')
    .option('--log-mode <normal|loose>', 'Log format mode.')
    .option(
      '--log-flight <[[no-]<date|title|colorful|inline>]>',
      'Enable / disable logger flights.',
      (val: string, acc: string[]) => acc.concat(val),
      [],
    )

    // Other builtin options.
    .option(
      '-c, --config-path, --configPath <configPath>',
      'config filepaths',
      (val, acc: string[]) => acc.concat(val),
      [],
    )
    .option(
      '--parastic-config-path, --parasticConfigFilepath <parasticConfigFilepath>',
      'parastic config filepath',
    )
    .option(
      '--parastic-config-entry, --parasticConfigEntry <parasticConfigEntry>',
      'parastic config filepath',
    )
    .option(
      '--workspace <workspace>',
      'The root dir to locate the config files and resolve relative paths',
    )

  return program
}
