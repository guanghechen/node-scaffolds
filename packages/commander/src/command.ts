import type { ICommandConfigurationOptions } from '@guanghechen/cli'
import type { OptionValues } from 'commander'
import { Command as Command$ } from 'commander'

/**
 * Callback for handling the command
 *
 * @param args    command arguments
 * @param options command options
 * @param extra   extra args (neither declared command arguments nor command options)
 * @param self    current command
 */
export type ICommandActionCallback<T extends ICommandConfigurationOptions> = (
  args: string[],
  options: T,
  extra: string[],
  self: Command,
) => void | Promise<void> | never

export class Command extends Command$ {
  // add missing type declarations
  protected _actionHandler: ((args: string[]) => void) | null | undefined
  protected _optionValues: object
  protected _storeOptionsAsProperties: boolean
  protected _version: string | undefined
  protected _versionOptionName: string | undefined

  constructor() {
    super()

    this._optionValues = {}
    this._storeOptionsAsProperties = false
  }

  /**
   * Register callback `fn` for the command.
   *
   * @example
   * program
   *   .command('serve')
   *   .description('start service')
   *   .action(function() {
   *      // do work here
   *   });
   *
   * @param {Function} fn
   * @return {Command} `this` command for chaining
   */
  public override action<T extends ICommandConfigurationOptions>(
    fn: ICommandActionCallback<T>,
  ): this {
    const listener = (args: string[]): unknown | Promise<unknown> => {
      // The .action callback takes an extra parameter which is the command or options.
      const expectedArgsCount = this.registeredArguments.length

      const actionArgs: Parameters<ICommandActionCallback<T>> = [
        args.slice(0, expectedArgsCount),
        this.opts() as T,
        args.slice(expectedArgsCount),
        this,
      ]

      const actionResult = fn.apply(this, actionArgs)
      return actionResult
    }

    this._actionHandler = listener
    return this
  }

  public override opts<T extends OptionValues>(): T {
    const nodes: Command[] = [this]
    for (let parent = this.parent; parent != null; parent = parent.parent) {
      nodes.push(parent as Command)
    }

    const options: Record<string, unknown> = {}
    for (let i = nodes.length - 1; i >= 0; --i) {
      const o = nodes[i]
      if (o._storeOptionsAsProperties) {
        for (const option of o.options!) {
          const key = option.attributeName()
          options[key] = key === o._versionOptionName ? o._version : (o as any)[key]
        }
      } else {
        const optionValues: Record<string, unknown> = o._optionValues as any
        for (const key of Object.getOwnPropertyNames(optionValues)) {
          options[key] = optionValues[key]
        }
      }
    }
    return options as unknown as T
  }
}
