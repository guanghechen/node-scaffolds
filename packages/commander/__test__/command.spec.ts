// @ts-ignore
import { chalk } from '@guanghechen/chalk/node'
import type {
  ICommandConfigurationFlatOpts,
  ICommandConfigurationOptions,
  IResolveCommandConfigurationOptionsParams,
} from '@guanghechen/cli'
import { resolveCommandConfigurationOptions } from '@guanghechen/cli'
import { createReporterMock } from '@guanghechen/helper-jest'
import { pathResolver } from '@guanghechen/path'
import type { IReporter } from '@guanghechen/reporter'
import { Reporter } from '@guanghechen/reporter'
import { convertToBoolean, cover, isNonBlankString } from '@guanghechen/std'
import { desensitize } from 'jest.helper'
import path from 'node:path'
import type { Command } from '../src'
import { createTopCommand } from '../src'

describe('command', () => {
  describe('no sub-command', () => {
    test('with arguments', async () => {
      const argv: string[] = [
        'node',
        'guanghechen',
        'source content waw waw waw',
        'this arg should be ignored',
        '--log-level=debug',
        '--log-flight=no-date',
        '--log-flight=no-colorful',
        '-f',
        '--no-silence',
      ]

      const reporter = new Reporter(
        chalk,
        {
          baseName: 'ghc',
          flights: { colorful: false, date: false },
        },
        argv,
      )
      const mock = createReporterMock({ reporter, desensitize })
      const result0 = await getCommand(argv, reporter)
      expect(desensitize(result0.args)).toMatchSnapshot('args')
      expect(desensitize(result0.options)).toMatchSnapshot('options')
      expect(desensitize(result0.resolvedOptions)).toMatchSnapshot('resolvedOptions')
      expect(mock.getIndiscriminateAll()).toMatchSnapshot('logs')
      mock.restore()
    })

    test('without arguments', async () => {
      const argv: string[] = [
        'node',
        'guanghechen',
        '--log-level=debug',
        '--log-flight=no-date',
        '--log-flight=no-colorful',
        '-f',
        '--no-silence',
      ]

      const reporter = new Reporter(
        chalk,
        {
          baseName: 'ghc',
          flights: { colorful: false, date: false },
        },
        argv,
      )
      const mock = createReporterMock({ reporter, desensitize })
      const result0 = await getCommand(argv, reporter)
      expect(desensitize(result0.args)).toMatchSnapshot('args')
      expect(desensitize(result0.options)).toMatchSnapshot('options')
      expect(desensitize(result0.resolvedOptions)).toMatchSnapshot('resolvedOptions')
      expect(mock.getIndiscriminateAll()).toMatchSnapshot('logs')
      mock.restore()
    })
  })
})

async function getCommand(
  argv: string[],
  reporter: IReporter,
): Promise<{
  args: string[]
  options: IGlobalCommandOptions
  resolvedOptions: IGlobalCommandOptions & ICommandConfigurationFlatOpts
  program: Command
}> {
  type ICommandData = Awaited<ReturnType<typeof getCommand>>

  const __defaultGlobalCommandOptions: IGlobalCommandOptions = {
    encoding: 'utf8',
    input: undefined,
    output: undefined,
    force: false,
    silence: true,
  }

  const program = createTopCommand('guanghechen', '0.0.0')
  const result: ICommandData = await new Promise<ICommandData>((resolve, reject) => {
    program
      .argument('[source content')
      .option('-e, --encoding <encoding>', 'Encoding of content from stdin or file.')
      .option('-i, --input <filepath>', 'Copy the data from <filepath> to the system clipboard.')
      .option(
        '-o, --output <filepath>',
        'Write the data from the system clipboard into <filepath>.',
      )
      .option('-f, --force', 'Overwrite the <filepath> without confirmation.')
      .option('-S, --no-silence', 'print info level logs.')
      .action(function (args: string[], options: IGlobalCommandOptions): void {
        try {
          const resolvedOptions = resolveGlobalCommandOptions<IGlobalCommandOptions>({
            commandName: '@guanghechen/commander--demo',
            defaultOptions: __defaultGlobalCommandOptions,
            reporter,
            options,
            subCommandName: '',
            workspace: path.resolve(),
          })
          resolve({ args, options, resolvedOptions, program })
        } catch (error) {
          reject(error)
        }
      })
      .parse(argv)
  })
  return result
}

interface IGlobalCommandOptions extends ICommandConfigurationOptions {
  encoding: BufferEncoding
  input?: string
  output?: string
  force: boolean
  silence: boolean
}

function resolveGlobalCommandOptions<O extends ICommandConfigurationOptions>(
  params: IResolveCommandConfigurationOptionsParams<O & IGlobalCommandOptions>,
): O & IGlobalCommandOptions & ICommandConfigurationFlatOpts {
  type R = O & IGlobalCommandOptions & ICommandConfigurationFlatOpts
  const baseOptions: R = resolveCommandConfigurationOptions<O & IGlobalCommandOptions>(params)
  const { reporter, options } = params

  // Resolve `encoding`.
  const encoding: string = cover<string>(baseOptions.encoding, options.encoding, isNonBlankString)
  reporter.debug('encoding:', encoding)

  // Resolve `input`.
  const _input: string | undefined = cover<string | undefined>(
    baseOptions.input,
    options.input,
    isNonBlankString,
  )
  const input: string | undefined = _input
    ? pathResolver.safeResolve(baseOptions.workspace, _input)
    : undefined
  reporter.debug('input:', input)

  // Resolve `output`.
  const _output: string | undefined = cover<string | undefined>(
    baseOptions.output,
    options.output,
    isNonBlankString,
  )
  const output: string | undefined = _output
    ? pathResolver.safeResolve(baseOptions.workspace, _output)
    : undefined
  reporter.debug('output:', output)

  // Resolve `force`.
  const force: boolean = cover<boolean>(baseOptions.force, convertToBoolean(options.force))
  reporter.debug('force:', force)

  // Resolve `silence`.
  const silence: boolean = cover<boolean>(baseOptions.silence, convertToBoolean(options.silence))
  reporter.debug('silence:', silence)

  return {
    ...baseOptions,
    encoding,
    input,
    output,
    force,
    silence,
  }
}
