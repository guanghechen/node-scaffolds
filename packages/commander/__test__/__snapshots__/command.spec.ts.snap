// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`command no sub-command with arguments: args 1`] = `
[
  "source content waw waw waw",
]
`;

exports[`command no sub-command with arguments: logs 1`] = `
[
  [
    "debug [ghc] cwd: <$WORKSPACE$>/packages/commander
",
  ],
  [
    "debug [ghc] workspace: <$WORKSPACE$>/packages/commander
",
  ],
  [
    "debug [ghc] configPath: []
",
  ],
  [
    "debug [ghc] parasticConfigPath: <$WORKSPACE$>/packages/commander/package.json
",
  ],
  [
    "debug [ghc] parasticConfigEntry: @guanghechen/commander--demo
",
  ],
  [
    "debug [ghc] logLevel: 1
",
  ],
  [
    "debug [ghc] resolvedConfig: {}
",
  ],
  [
    "debug [ghc] resolvedOptions: {encoding:'utf8',force:false,silence:true}
",
  ],
  [
    "debug [ghc] encoding: utf8
",
  ],
  [
    "debug [ghc] input: undefined
",
  ],
  [
    "debug [ghc] output: undefined
",
  ],
  [
    "debug [ghc] force: true
",
  ],
  [
    "debug [ghc] silence: false
",
  ],
]
`;

exports[`command no sub-command with arguments: options 1`] = `
{
  "configPath": [],
  "force": true,
  "logFlight": [
    "no-date",
    "no-colorful",
  ],
  "logLevel": "debug",
  "silence": false,
}
`;

exports[`command no sub-command with arguments: resolvedOptions 1`] = `
{
  "configPath": [],
  "cwd": "<$WORKSPACE$>/packages/commander",
  "encoding": "utf8",
  "force": true,
  "input": undefined,
  "logLevel": 1,
  "output": undefined,
  "parasticConfigEntry": "@guanghechen/commander--demo",
  "parasticConfigPath": "<$WORKSPACE$>/packages/commander/package.json",
  "silence": false,
  "workspace": "<$WORKSPACE$>/packages/commander",
}
`;

exports[`command no sub-command without arguments: args 1`] = `
[
  undefined,
]
`;

exports[`command no sub-command without arguments: logs 1`] = `
[
  [
    "debug [ghc] cwd: <$WORKSPACE$>/packages/commander
",
  ],
  [
    "debug [ghc] workspace: <$WORKSPACE$>/packages/commander
",
  ],
  [
    "debug [ghc] configPath: []
",
  ],
  [
    "debug [ghc] parasticConfigPath: <$WORKSPACE$>/packages/commander/package.json
",
  ],
  [
    "debug [ghc] parasticConfigEntry: @guanghechen/commander--demo
",
  ],
  [
    "debug [ghc] logLevel: 1
",
  ],
  [
    "debug [ghc] resolvedConfig: {}
",
  ],
  [
    "debug [ghc] resolvedOptions: {encoding:'utf8',force:false,silence:true}
",
  ],
  [
    "debug [ghc] encoding: utf8
",
  ],
  [
    "debug [ghc] input: undefined
",
  ],
  [
    "debug [ghc] output: undefined
",
  ],
  [
    "debug [ghc] force: true
",
  ],
  [
    "debug [ghc] silence: false
",
  ],
]
`;

exports[`command no sub-command without arguments: options 1`] = `
{
  "configPath": [],
  "force": true,
  "logFlight": [
    "no-date",
    "no-colorful",
  ],
  "logLevel": "debug",
  "silence": false,
}
`;

exports[`command no sub-command without arguments: resolvedOptions 1`] = `
{
  "configPath": [],
  "cwd": "<$WORKSPACE$>/packages/commander",
  "encoding": "utf8",
  "force": true,
  "input": undefined,
  "logLevel": 1,
  "output": undefined,
  "parasticConfigEntry": "@guanghechen/commander--demo",
  "parasticConfigPath": "<$WORKSPACE$>/packages/commander/package.json",
  "silence": false,
  "workspace": "<$WORKSPACE$>/packages/commander",
}
`;
